ENH: --inherit for add/create-sibling, RF: publish --missing=fail|skip|inherit; wanted for annex; apparently some BF for ls

yarikoptic opened the pull request:
Closes #1295 

- [x] get feedback from fella developers
- [x] inherit annex wanted settings
- [x] may be move some of introduced code into functions to improve modularity etc

yarikoptic commented:
lovely tests catch awkward double `\\`'ing spaces within ssh RIs for the remote... diidn't figure out where it is coming from

yarikoptic commented:
@mih and @bpoldrack -- it is conceptually ready and seems to work. feedback?
will deal with failing tests if any tomorrow

yarikoptic commented:
BTW - as for wanted -- decided just to inherit them from super datasets which are annexes... which adds a cent to have supers annexes nearly always

bpoldrack commented on datalad/distribution/publish.py:
```diff
@@ -255,7 +256,10 @@ class Publish(Interface):
-        annex_copy_opts=annex_copy_opts
+        annex_copy_opts=annex_copy_opts,
+        # "interferes" with skip_failing which is actually "skip_missing"
+        # TODO: unify as --missing=(skip|inherit) ?
```
ATM it's `skip_missing` indeed. But `skip_failing` is also used, when there was no sibling specified and we can not figure out one automatically. In that case `inherit` wouldn't work, since we don't know what to inherit from. With proper documentation, `inherit` could then fall back to `skip` behavior, but we then also need a third value `fail`, I guess.

yarikoptic replied:
so, semantically it is "skip missing" not "failing" since not that some publish of a sub-dataset has failed (which can happen) and we would like to skip it then.
So do you think I should tackle RF `--skip-failing` + `--inherit-settings` into `--missing=(fail|skip|inherit)` within this PR?  I guess I could as well do it

bpoldrack replied:
Yes, makes sense to me.

bpoldrack commented on datalad/distribution/publish.py:
```diff
@@ -343,11 +348,25 @@ def __call__(
+                    # would not be as transparent to inherit for -github
+                    lgr.info("Will try to create a sibling inheriting settings from %s", superds)
+                    # XXX explicit None as sshurl for now
+                    ds.create_sibling(None, name=to, inherit_settings=True)
```
I'm sceptical. :-)
Shouldn't it be `ds.add_sibling(..., inherit_settings=True)`? Actually creating a sibling on the remote from within `publish` seems a little bit harsh to me. While it might make sense in some cases, it would require different access permissions than before in other cases. Also: What's the policy regarding parameter `existing` in that case?
I think inheriting settings for the remote configuration locally is fine, but creating a fitting remote repo is something different.

bpoldrack replied:
May be we just need more detailed values for `inherit` to distinguish whether we should try to create or just add that sibling.

yarikoptic replied:
overall idea here was to improve UX -- I got tired of remembering either I have new sub-datasets or not before running `publish`, and then invoking `create-sibling` manually in case if there were and publish failed...  I had to remember parameters for `create-sibling` invocation etc -- just doesn't scale.  Hence I have decided to address at least this common use-case where I have a hierarchy of datasets, which I could publish over and over again efficiently even if I get new sub-datasets without remembering again and again how to create those remote again and again.
Such scheme indeed might fail in case if there is two sets of hierarchies which try to sync/publish  to the common one -- that is where `existing=skip` would be necessary I guess.  For now `existing`  here indeed remained to its safest default and I did not proxied any more settings into `create_sibling` -- for any of those scenarios better to use it directly/consciously.
Having said all that I am yet to see this piece "in action" since yesterday had to revert to separate `create-sibling` invocation anyways and `publish` is hard to tame for me (see #1329) ATM

bpoldrack commented on datalad/support/annexrepo.py:
```diff
@@ -513,6 +513,31 @@ def _run_annex_command(self, annex_cmd, git_options=None, annex_options=None,
+        out, err = self._run_annex_command(
+            *args, **kwargs
+        )
+        lines = out.rstrip('\n').split('\n')
```
Nope. `'\n'` is platform dependent.
```python
lines = out.splitlines()
```

yarikoptic replied:
thanks!

yarikoptic commented on datalad/distribution/add_sibling.py:
```diff
@@ -298,6 +306,16 @@ def __call__(name=None, url=None, dataset=None,
+                    repo.set_group(name, annex_group)
+                if annex_groupwanted:
+                    if not annex_group:
+                        raise ValueError("To set groupwanted, you need to provide annex_group option")
```
thanks!

yarikoptic commented:
oki doki 

- publish now got --missing=fail|skip|inherit
- moved relevant inheriting logic into add-sibling, so it got --inherit too now.  Doesn't have a dedicated test though but test ran all the way from the top by publish should check majority of the conditioning ;)
- also renamed --inherit-settings into just --inherit to stay inline with `--missing=inherit`

so if tests pass and no objections -- wouldn't mind a merge. real-life testing would show more

yarikoptic commented:
for me few tests fail ATM due to change in publish behavior, but for CI might be just fine, so if tests pass and no other comments -- a candidate for merge

bpoldrack commented:
Will merge for now. See #1344 .
